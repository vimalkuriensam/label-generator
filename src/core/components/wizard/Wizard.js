import React, { Component } from "react";
import ShippingLabelMaker from "../../../features/shipping-label-maker/ShippingLabelMaker";

import Stepper from "../stepper/Stepper";

class Wizard extends Component {
  state = {
    steps: [
      "Sender Address",
      "Receiver Address",
      "Weight",
      "Option",
      "Confirm",
    ],
    currentStep: 1,
    stepCount: 1,
    endAction: false,
    paymentAmount: 0,
  };

  endAction = (values) => {
    this.onComplete(values);
  };

  onComplete = (values) => {
    const { weight, shippingRate, ground } = values;
    const paymentAmount =
      parseInt(weight) * shippingRate * (parseInt(ground) === 1 ? 1 : 1.5);
    this.setState(() => ({ paymentAmount }));
    this.setState(() => ({ endAction: true }));
    setTimeout(() => {
      this.setState(() => ({ endAction: false }));
      this.setState(() => ({ paymentAmount: 0 }));
      this.setState(() => ({ currentStep: 1 }));
      this.setState(() => ({ stepCount: 1 }));
    }, 5000);
  };

  onAction = (e, state = "forward") => {
    e.preventDefault();
    if (state === "forward") {
      this.setState((prevState) => ({
        currentStep: prevState.currentStep + 1,
      }));
      this.setState((prevState) => ({ stepCount: prevState.stepCount + 1 }));
    } else {
      this.setState((prevState) => ({
        currentStep: prevState.currentStep - 1,
      }));
      this.setState((prevState) => ({ stepCount: prevState.stepCount - 1 }));
    }
  };
  render() {
    return (
      <div className="wizard">
        <div className="wizard__stepper">
          <Stepper
            steps={this.state.steps}
            currentStep={this.state.currentStep}
          />
        </div>
        <div className="wizard__wrapper">
          {!this.state.endAction ? (
            <ShippingLabelMaker
              stepCount={this.state.stepCount}
              onAction={this.onAction}
              onComplete={this.endAction}
            />
          ) : (
            <div>
              <div className="row">
                <div className="col-1-of-2">Payment</div>
                <div className="col-1-of-2">{this.state.paymentAmount}</div>
              </div>
              <div
                style={{
                  fontWeight: "bold",
                  color: "black",
                  fontSize: "1.rem",
                  textAlign: "center",
                }}
              >
                Redirecting...
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }
}

export default Wizard;
