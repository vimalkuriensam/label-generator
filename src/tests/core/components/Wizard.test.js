import React from "react";
import { shallow } from "enzyme";

import Wizard from "../../../core/components/wizard/Wizard";

test("Should render wizard component correctly", () => {
  const wrapper = shallow(<Wizard />);
  expect(wrapper).toMatchSnapshot();
});

test("Should increase stepper value on onAction click for next", () => {
  const wrapper = shallow(<Wizard />);
  const currentStep = wrapper.state("currentStep");
  const stepCount = wrapper.state("stepCount");
  wrapper.find("ShippingLabelMaker").prop("onAction")(
    { preventDefault: () => {} },
    "forward"
  );
  expect(wrapper.state("currentStep")).toBe(currentStep + 1);
  expect(wrapper.state("stepCount")).toBe(stepCount + 1);
  expect(wrapper).toMatchSnapshot();
});

test("Should decrease stepper value on onAction click for previous", () => {
  const wrapper = shallow(<Wizard />);
  const currentStep = wrapper.state("currentStep");
  const stepCount = wrapper.state("stepCount");
  wrapper.find("ShippingLabelMaker").prop("onAction")(
    { preventDefault: () => {} },
    "backward"
  );
  expect(wrapper.state("currentStep")).toBe(currentStep - 1);
  expect(wrapper.state("stepCount")).toBe(stepCount - 1);
  expect(wrapper).toMatchSnapshot();
});

test("Should call onComplete function when endAction is called from shippingLabelMaker", () => {
  const wrapper = shallow(<Wizard />);
  wrapper.find('ShippingLabelMaker').prop('onComplete')();
  expect(wrapper.state('endAction')).toBe(true);
  expect(wrapper).toMatchSnapshot();
})

test("Should printout shipping cost correctly", () => {
  const value = {
    weight: 1,
    shippingRate: 0.4,
    ground: 1,
  }
  const wrapper = shallow(<Wizard />);
  wrapper.find('ShippingLabelMaker').prop('onComplete')(value);
  expect(wrapper.state('paymentAmount')).toBe(0.4);
})
