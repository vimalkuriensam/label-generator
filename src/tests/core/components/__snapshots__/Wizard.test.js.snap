// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render wizard component correctly 1`] = `
<div
  className="wizard"
>
  <div
    className="wizard__stepper"
  >
    <Stepper
      currentStep={1}
      steps={
        Array [
          "Sender Address",
          "Receiver Address",
          "Weight",
          "Option",
          "Confirm",
        ]
      }
    />
  </div>
  <div
    className="wizard__wrapper"
  >
    <withRouter(ShippingLabelMaker)
      onAction={[Function]}
      onComplete={[Function]}
      stepCount={1}
    />
  </div>
</div>
`;
